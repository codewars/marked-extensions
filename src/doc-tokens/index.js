import { replaceDocTypes } from './doc-types';
import { replaceDocNames } from './doc-names';
import { replaceDocGlobals } from './doc-globals';

/**
 * Preprocesses the markdown before sending it through marked. This is used to process
 * doc tokens and any other future extensions that we support that don't require being handled
 * during the marked rendering.
 * @param result
 */
export function processDocTokens(result, html, pre) {
  const language = result.originalLanguage || result.language;
  const globals = replaceDocGlobals(language, pre, html);
  const types = replaceDocTypes(language, pre, globals);
  const names = replaceDocNames(language, pre, types);

  return names;
}
