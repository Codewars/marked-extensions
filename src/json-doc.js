export function jsonDoc(code) {
  try {
    let json = JSON.parse(code);
    const md = [methodName(json)];
    if (json.desc) {
      md.push(json.desc);
    }
    md.push('```%doc\nParameters:');
    md.push(parameters(json));
    md.push('Return Value:');
    md.push(returnType(json));
    if (json.constraints && json.constraints.length) {
      md.push('Constraints:');
      md.push(json.constraints.join('\n'));
    }
    md.push('```');
    if (json.examples && json.examples.length) {
      md.push('```%doc-block');
      md.push('#### Examples');
      md.push(exampleHeader(json));
      md.push(exampleRows(json));
      md.push('```');
    }

    return md.join('\n');
  }
  catch (ex) {
    return '`Failed to render %jsonblock: ' + ex.message + '`';
  }
}

function exampleRows(json) {
  return json.examples.map(exampleRow).join('\n');
}

function exampleRow(example, index) {
  const name = example.name || `Example #${index+1}`;
  let md = `**${name}**|`;
  md += example.args.map(arg => JSON.stringify(arg)).join('|');
  md += `|${example.returns || ''}`;
  return md;
}

function exampleHeader(json) {
  const line1 = [''];
  const line2 = [''];
  Object.keys(getArgs(json)).forEach(key => {
    line1.push(key)
    line2.push('');
  });
  line1.push('returns');
  return `${line1.join('|')}\n-${line2.join('|-')}|-`;
}

function getArgs(json) {
  return json.args || json.params || json.parameters || {};
}

function methodName(json) {
  const args = Object.keys(getArgs(json)).map(key => `\`@@docName:${key}\``);
  return `### \`@@docGlobal:Challenge.@@docMethod:${json.method}\`(${args.join(', ')})`;
}

function parameters(json) {
  const args = getArgs(json);
  const params = Object.keys(args).map(key => {
    const arg = args[key];
    const type = typeof arg === 'string' ? arg : arg.type;
    let md = `@@docName:${key}: @@docType:${type || 'String'}`;
    if (arg.desc) {
      md += ` - ${arg.desc}`;
    }

    return md;
  });

  return params.join('\n');
}

function returnType(json) {
  if (json.returns) {
    const type = typeof json.returns === 'string' ? json.returns : json.returns.type;
    let md = `@@docType:${type || 'void'}`;
    if (json.returns.desc) {
      md += ` - ${json.returns.desc}`;
    }

    return md;
  }
  return '@@docType:void';
}
