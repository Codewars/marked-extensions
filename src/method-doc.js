export function methodDoc(code) {
  try {
    let json = JSON.parse(code);
    const md = [methodHeader(json)];
    if (json.desc) {
      md.push(json.desc);
    }
    md.push('```%doc');

    if (json.args) {
      md.push('Parameters:');
      md.push(parameters(json));
    }
    md.push('Return Value:');
    md.push(returnType(json));
    if (json.constraints && json.constraints.length) {
      md.push('Constraints:');
      md.push(json.constraints.join('\n'));
    }
    if (json.errors && json.errors.length) {
      md.push('Errors:');
      md.push(json.errors.join('\n'));
    }
    md.push('```');
    if (json.examples && json.examples.length) {
      md.push('```%doc-block');
      md.push('#### Examples');
      md.push(exampleHeader(json));
      md.push(exampleRows(json));
      md.push('```');
    }

    return md.join('\n');
  }
  catch (ex) {
    return '`Failed to render %jsonblock: ' + ex.message + '`';
  }
}

function exampleRows(json) {
  return json.examples.map((v, i) => exampleRow(json, v, i)).join('\n');
}

function exampleRow(json, example, index) {
  const name = example.name || `Ex. #${index+1}`;
  let md = `*${name}*|`;
  if (example.args) {
    md += example.args.map(arg => '`' + JSON.stringify(arg) + '`').join('|');
  }
  md += `|\`${JSON.stringify(example.returns) || ''}\``;
  return md;
}

function exampleHeader(json) {
  const line1 = [''];
  const line2 = [''];
  Object.keys(getArgs(json)).forEach(key => {
    line1.push(key)
    line2.push('');
  });
  line1.push('Return Value');
  return `${line1.join('|')}\n-${line2.join('|-')}|-`;
}

function getArgs(json) {
  return json.args || json.params || json.parameters || {};
}

function methodName(json) {
  let globalName = json.global !== false ? `@@docGlobal:${json.global || 'Challenge'}.` : '';
  // if a class is provided, it will always be shown and overrides global
  if (json.class) {
    globalName = `@@docClass:${json.class}.`;
  }

  return `${globalName}@@docMethod:${json.method}`;
}

function methodHeader(json) {
  const args = Object.keys(getArgs(json)).map(key => `\`@@docName:${key}\``);
  return `### \`${methodName(json)}\`(${args.join(', ')})`;
}

function parameters(json) {
  const args = getArgs(json);
  const params = Object.keys(args).map(key => {
    const arg = args[key];
    const type = typeof arg === 'string' ? arg : arg.type;
    let md = `@@docName:${key}: @@docType:${type || 'String'}`;
    if (arg.desc) {
      md += ` - ${arg.desc}`;
    }

    return md;
  });

  return params.join('\n');
}

function returnType(json) {
  if (json.returns) {
    const type = typeof json.returns === 'string' ? json.returns : json.returns.type;
    let md = `@@docType:${type || 'void'}`;
    if (json.returns.desc) {
      md += ` - ${json.returns.desc}`;
    }

    return md;
  }
  return '@@docType:void';
}
