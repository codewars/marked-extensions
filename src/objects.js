/**
 * Assigns the keys from source into target, but only those that are not already assigned. Useful
 * in place of Object.assign which would require a polyfill
 * @param target
 * @param source
 * @returns {*}
 */
export function assignMissing(target, source) {
  Object.keys(source || {}).forEach((key) => {
    if (!(key in target)) {
      target[key] = source[key];
    }
  });
  return target;
}
