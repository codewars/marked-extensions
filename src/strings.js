export function upperCaseWords(str) {
  if (!str) return str;
  return str.replace(/(?:^|\s)\S/g, (a) => a.toUpperCase());
}

export function camelCase(str, pascal) {
  const cased = upperCaseWords(str.replace(/_/g, ' ')).replace(/ /g, '');
  return pascal ? cased : lowerCaseFirst(cased);
}

export function lowerCaseFirst(str) {
  return `${str[0].toLowerCase()}${str.substr(1)}`;
}

export function escapeHtml(text) {
  return text ? text.replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;') : text;
}

export function unescapeHtml(text) {
  return text ? text.replace(/&amp;/g, '&').replace(/&lt;/g, '<').replace(/&gt;/g, '>') : text;
}
