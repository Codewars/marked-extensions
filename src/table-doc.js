export function tableDoc(code) {
  try {
    let json = JSON.parse(code);
    const html = [];

    if (json.table) {
      html.push(`<h3>${json.table}</h3>`);
    }

    if (json.desc) {
      html.push(`<p>${json.desc}</p>`);
    }

    html.push('<table>');
    html.push('<tr><th>Name</th><th>Type</th></tr>');

    if (json.columns){
      html.push(tableHeaders(json.columns));
    }

    html.push('</table>');
    return html.join('\n');
  }
  catch (ex) {
    return '`failed to render %jsonblock: `' + ex.message + '`';
  }
}

const tableHeaders = json =>
  Object.keys(json)
  .map(k => `<tr><td>${k}</td><td>${json[k]}</td></tr>`)
  .join('\n');
