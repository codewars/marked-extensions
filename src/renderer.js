import { assignMissing } from './objects'
import { processDocTokens } from './doc-tokens'
import { methodDoc } from './method-doc';
import { tableDoc } from './table-doc';



export function buildRenderer(marked, options, result) {
  const renderer = result.renderer = new marked.Renderer();

  const markedOptions = { renderer };
  assignMissing(markedOptions, options.marked || {});

  // provide the render method, this will also be used later to render nested blocks
  result.render = (md) => marked(md, markedOptions);

  setupHeader(renderer, options, result);
  setupCode(options, result);
}

/**
 *
 * @param renderer
 */
function setupHeader(renderer, options, result) {
  // heading extensions
  renderer.heading = function (text, level) {
    // you can set icons via icon::name
    const icon = text.match(/icon::([a-z-]*)/);
    let attributes = '';
    if (icon) {
      // indicate that this icon has been used
      result.icons[icon[1]] = true;

      attributes = ` class="${options.iconClassPrefix}${icon[1]}"`;
      text = text.replace(/icon::[a-z-]*/, '');
    }

    // we track headers 1 - 4 and add ids to them so that we can link to them later if we want to
    if (level < 5) {
      const header = result.headers[`h${level}`];
      const index = header.length;
      attributes += ` id="h${level}_${index}"`;
      header.push(text);
    }

    return `<h${level}${attributes}>${text}</h${level}>`;
  }
}

/**
 * Handles code blocks in a variety of ways
 * @param options
 * @param result
 */
function setupCode(options, result) {
  const _code = result.renderer.code;

  // special version of render that will process doc tokens. Needed at times when
  // tokens are nested inside of pre and should be rendered without tags (but their labels should still be processed)
  const render = (code, preTokens) => {
    if (preTokens && options.docTokens) {
      code = processDocTokens(result, code, true);
    }

    return result.render(code);
  }


  result.renderer.code = function(code, language) {
    if (language) {
      if (language.match(/^if:/)) {
        return matchIfBlockLanguage(result, language) ? render(code) : '';
      }
      else if (language.match(/^if-not:/)) {
        return matchIfBlockLanguage(result, language) ? '' : render(code);
      }
      else if (language.match(/^tab:/)) {
        return handleTab(result, code, language);
      }
      else if (result.extensions.indexOf(language) >= 0) {
        return handleExtension(options, result, code, language);
      }
      else if (language === '%definitions' || language === '%doc') {
        return wrapInBlockDiv(language, renderDefinitions(result, code, render));
      }
      else if (language === '%method-doc') {
        return wrapInBlockDiv('docs method-doc', render(methodDoc(code, result.originalLanguage)));
      }
      else if(language === '%table-doc') {
        return wrapInBlockDiv('docs table-doc', tableDoc(code))
     }
      else if (language[0] === '%') {
        return wrapInBlockDiv(language, result.render(code));
      } 

      // make sure this is a language and not some random tag
      const foundLanguage = options.findLanguage(language.split(':')[0]);

      if (foundLanguage) {
        // if filtering is enabled and this is not the active language then filter it out
        if (options.filterLanguages && foundLanguage !== result.language && result.language) {
          return '';
        }

        // if CodeMirror is provided then highlight using that instead
        if (options.cm) {
          return highlightCM(options, code, foundLanguage, language);
        }
      }
      // process line numbers, if they are set (i.e. ruby:10) in a naive simple way
      else if (options.lineNumbers) {
        code = lineNumbers(code, language);
      }
    }

    return wrapLanguage(options, _code.call(result.renderer, code, language), language);
  }
}

function lineNumbers(code, language, wrapper = '@@ ') {
  let lineNumber = getLineNumber(language);

  // if there are line numbers, then add them now starting at the start index
  if (lineNumber > 0) {
    code = code.split('\n').map(line => {
      return `${wrapper.replace('@@', lineNumber++)}${line}`
    }).join('\n');
  }

  return code;
}

function wrapLanguage(options, code, language) {
  // if we have reached this point then CM isn't enabled and we need to
  if (language && options.languageWrapper) {
    if (typeof options.languageWrapper === 'function') {
      code = options.languageWrapper(code, language);
    }
    else {
      code = options.languageWrapper.replace('{slot}', code);
    }
  }

  return code;
}

function highlightCM(options, code, language, raw) {
  const el = window.document.createElement('div');
  options.cm.runMode(code, options.findMode(language), el);

  let codeHtml = el.innerHTML;
  if (options.lineNumbers) {
    codeHtml = lineNumbers(codeHtml, raw, '<span class="cm-line-number">@@</span>');
  }

  const result = `<pre class="cm-runmode cm-s-${options.theme}"><code>${codeHtml}</code></pre>`;
  return wrapLanguage(options, result, language);
}

function getLineNumber(language) {
  const parts = language.split(':');
  return parts.length > 1 ? parseInt(parts[1], 10) : null;
}

function wrapInBlockDiv(type, contents) {
  return `<div class="block block--${type.replace(/^%/, '')}">${contents}</div>`;
}

function matchIfBlockLanguage(result, language) {
  return language.replace(/^if(-not)?: ?/, '').split(',').indexOf(result.originalLanguage) >= 0;
}

function handleTab(result, code, language) {
  // parts should be up to tab:LABEL with language being optional
  const parts = language.split(':');
  let label = parts[1].replace(/\+/g, ' ');
  result.tabs[label] = `${result.render(code)}`;
  return '';
}

/**
 * If the extension value is a function, it will treat it as a render function, otherwise it will
 * assume the extension value is a string and treat it as a template with {code} as the code placeholder.
 * @param options
 * @param code
 * @param language
 */
function handleExtension(options, result, code, language) {
  const ext = options.extensions[language]

  if (typeof ext.code === 'function') {
    return ext.code(code, options);
  }
  else {
    return ext.code.replace('{code}', code);
  }
}

/**
 *
 * @param result
 * @param code
 * @returns {string}
 */
function renderDefinitions (result, code, render) {
  var html = '<dl>';
  if (code) {
    code.split('\n').forEach(line => {
      if (line.match(/^#/)) {
        html += render(line);
      }
      else if (line.match(/: *$/)) {
        html += `<dt>${line.replace(/:$/, '')}</dt>`;
      }
      else {
        // if line starts with 4 white spaces, a tab or a ` block, then consider it a pre tag and don't render dfn
        html += `<dd>${render(line, !!line.match(/^(\t|\s{4}|`)/))}</dd>`;
      }
    });
  }

  return html + '</dl>';
}
