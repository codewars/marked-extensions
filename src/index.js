import { assignMissing } from './objects'
import { buildRenderer } from './renderer'
import { processDocTokens } from './doc-tokens'

export const defaultOptions = {
  // these are the options passed to marked directly
  marked: {},
  // languages that should be treated as extensions. You can configure how each gets handled
  extensions: {
    mermaid: {
      // code handler
      code: code => `<div class="mermaid">${code}</div>`,
      // will lazy load script automatically
      src: 'https://cdn.rawgit.com/knsv/mermaid/6.0.0/dist/mermaid.min.js',
      afterRender: () => window.mermaid && window.mermaid.init()
    },
  },
  // set to a method that that will receive a language and return the mapped
  // name. if no name is returned, it is assumed that the language is an extension.
  // The default method only handles a default set of languages.
  findLanguage: (language) => defaultLanguages.indexOf(language) >= 0 ? language: null,

  // used with CodeMirror.runMode, used to find the mode of the language. By default will just return the language
  findMode: (language) => language,

  // The highlight theme to add, only used with cm runMode
  theme: 'neo',

  // you can set icons within headers using icon::ICONNAME, this setting determines the icon class prefix used
  iconClassPrefix: 'icon-',

  // set to a value that should wrap standard languages. Use "{slot}" to indicate where the code should be inserted: i.e.
  // '<div class="tab">{code}</div>'. Can also be a function which takes (code, language) as its parameters.
  languageWrapper: null,

  // the language that should be used for filtering out all other languages
  language: null,

  // true if languages should be filtered
  filterLanguages: true,

  // if the language value is not set, or is set to a value not found within the markdown, this
  // setting will default the language to the first one found within the markdown.
  defaultLanguageToFirst: true,

  // true if line numbers should be added to highlighted code (if a starting line number is configured for the block)
  lineNumbers: true,

  // css class used to display a line number gutter, which can be used to try to create an opacity layer
  // which will style the cm-number values differently. Set to null to disable adding a gutter.
  lineNumbersGutter: 'cm-runmode-linenumbers-gutter',

  // true if doc tokens should be parsed
  docTokens: true,

  // if set to an instance of js-yaml, it will process meta data at the top of the markdown.
  jsYaml: null,

  // If set to CodeMirror class, it will use CodeMirror.runMode to process syntax highlighting
  // Note: make sure to import codemirror/addon/runmode/runmode.js first.
  cm: null,

  // set to automatically load a CM language. This is provided by default but you can override if you wish
  // to override. This option is only used if both cm and loadScript are set.
  // You can set this to null if you do not wish or need to load languages dynamically
  loadCMLanguage: (language, options) => {
    return options.loadScript(`//cdnjs.cloudflare.com/ajax/libs/codemirror/${options.cm.version}/mode/${language}/${language}.min.js`);
  },

  // If you wish to support loading external extension scripts, you should set this to a
  // function that takes a url and returns a promise. Note that this function will need to be responsible
  // for not reloading the same scripts if requested more than once, this library does not take care of caching.
  loadScript: null,

  // if set to a function, will be called back after all external scripts have loaded
  onLoaded: null
};

export const defaultLanguages = [
  'c', 'clojure', 'coffeescript', 'cpp', 'csharp', 'elixir', 'erlang', 'fsharp',
  'go', 'groovy', 'haskell', 'java', 'javascript', 'kotlin', 'objc', 'ocaml', 'php', 'python',
  'r', 'ruby', 'scala', 'shell', 'solidity', 'sql', 'swift', 'typescript'
]

/**
 * Processes the markdown using marked along with the many extensions this library provides
 * @param marked The marked library, must be passed in since it is not included within this library as a dependency
 * @param markdown The markdown to process
 * @param options The extended set of options, as well as marked options. See defaultOptions for more details.
 * @returns {{originalLanguage, language, languages: [], extensions: [], tabs: {}, headers: {h1: Array, h2: Array, h3: Array, h4: Array}, icons: [], raw: *, preprocessed: *}}
 */
export function process(marked, markdown, options = {}) {
  assignMissing(options, defaultOptions);

  const result = {
    originalLanguage: options.language,
    language: options.language,
    languages: {},
    extensions: {},
    tabs: {},
    headers: {h1: [], h2: [], h3: [], h4: []},
    icons: {},
    raw: markdown,
    preprocessed: markdown
  };

  if (options.jsYaml) {
    processMeta(options, result);
  }

  processBlocks(options, result);
  if (options.defaultLanguageToFirst) {
    processLanguage(result);
  }

  buildRenderer(marked, options, result);

  var html = null;
  result.html = () => html || (html = render(options, result));
  result.afterRender = afterRenderFn(options, result);

  // convert objects which have been acting as basic sets to an array
  ['languages', 'extensions', 'icons']
    .forEach(key => result[key] = Object.keys(result[key]));

  if (options.loadScript) {
    processExternalScripts(options, result);
  }

  return result;
}

function render(options, result) {
  let html = result.render(result.preprocessed);

  if (options.docTokens) {
    html = processDocTokens(result, html);
  }

  return html;
}

/**
 * Creates the afterRender function that is added to the result, which can be called once the
 * processed html has been added to the DOM to initialize any extensions that may have been loaded.
 * @param options
 * @param result
 * @returns {Function}
 */
function afterRenderFn(options, result) {
  return function() {
    result.extensions.forEach(ext => {
      const config = options.extensions[ext];
      if (config && config.afterRender) {
        config.afterRender.apply(result, arguments);
      }
    });
  }
}

/**
 * Will loop through extensions and languages and try to dynamically load scripts.
 * @param options
 * @param result
 */
function processExternalScripts(options, result) {
  const promises = [];
  result.extensions.forEach(ext => {
    const config = options.extensions[ext];
    if (config && config.src) {
      promises.push(options.loadScript(config.src));
    }
  });

  // automatically load the CM language that is shown
  if (options.cm && options.loadCMLanguage) {
    // if we filter languages then we only need to load the one being shown
    if (options.filterLanguages && result.language) {
      promises.push(options.loadCMLanguage(result.language, options));
    }
    // otherwise we need to load them all
    else if (!options.filterLanguages && result.languages.length) {
      result.languages.forEach(language => {
        promises.push(options.loadCMLanguage(language, options));
      });
    }
  }

  Promise.all(promises).then(() => {
    if (options.onLoaded) {
      options.onLoaded(result);
    }
  });
}

/**
 * Processes yaml content at the top of the markdown, marked starting by a --- and ending with a ...
 * @param options
 * @param result
 */
function processMeta(options, result) {
  result.preprocessed = result.preprocessed.replace(/^---\r?\n(.*\r?\n)*\.{3}\s*(\r?\n){2}?/, meta => {
    let yaml = meta.replace(/^---\r?\n/, '').replace(/\r?\n\.\.\. *\r?\n?/, '');
    result.meta = options.jsYaml.safeLoad(yaml);
    return '';
  });
}

/**
 * if no language was provided, or the one provided is not in the list of supported languages,
 * then switch to the first language found
 * @param result
 */
function processLanguage(result) {
  if (!result.language || !result.languages[result.language]) {
    result.language = Object.keys(result.languages)[0];
  }
}

/**
 * Loops through all ``` style blocks and figures out which are languages and which are
 * possibly extensions
 * @param markdown
 * @param options
 * @param result
 */
function processBlocks(options, result) {
  let blocks = result.preprocessed.match(/^(```|~~~) ?(.*) *$/gm) || [];
  blocks = blocks.map(m => m.replace(/(```|~~~) ?/g, ''));

  // loop through each block and track which are languages and which are extensions
  blocks.forEach((text) => {
    if (text) {
      text = text.replace(/^if(-not)?: ?/, '').split(':')[0];

      text.split(',').forEach(name => {
        // % is a special token, we know these aren't either languages or extensions
        if (name.indexOf('%') === -1) {
          // if an extension has been defined for the language, track it now
          if (options.extensions[name]) {
            result.extensions[name] = true;
          }
          else {
            const language = options.findLanguage(name);
            if (language) {
              result.languages[language] = true;
            }
          }
        }
      })
    }
  });
}
