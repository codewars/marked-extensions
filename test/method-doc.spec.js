import { process } from '../src/index';
import { marked } from 'marked';
import { fixture } from './test-utils';

describe('%method-doc', () => {
  it('should render json as formatted markdown', () => {
    const example = process(marked, fixture('method-doc')).html();
    expect(example).toEqual(expect.stringContaining('>Examples</h4>'));
    expect(example).toEqual(expect.stringContaining('<th>files</th>'));
    // Used to have empty `<dfn class="doc-class"></dfn>` after global prefix was removed
    // expect(example).toEqual(expect.stringContaining('<code><dfn class="doc-class">'));
    expect(example).toEqual(
      expect.stringContaining('- A filtered array of files that were opened</p>')
    );
    expect(example).toEqual(expect.stringContaining('<td><code>[1,2,3]</code></td>'));
    expect(example).toEqual(expect.stringContaining('<td><code>"js"</code></td>'));
    expect(example).toEqual(expect.stringContaining('<table>'));
  });

  it('should render json as formatted markdown', () => {
    const example = process(marked, fixture('method-doc-invalid')).html();
    expect(example).not.toEqual(expect.stringContaining('>Examples</h4>'));
    expect(example).toEqual(expect.stringContaining('openFiles'));
  });

  it('should match snapshot', () => {
    // Uses https://github.com/algolia/jest-serializer-html to make the snapshot diff readable
    const example = process(marked, fixture('method-doc')).html();
    expect(example).toMatchSnapshot();
  });

  it('should match snapshot with language: javascript', () => {
    const example = process(marked, fixture('method-doc'), { language: 'javascript' }).html();
    expect(example).toMatchSnapshot();
  });

  it('should match snapshot with language: csharp', () => {
    const example = process(marked, fixture('method-doc'), { language: 'csharp' }).html();
    expect(example).toMatchSnapshot();
  });
});
