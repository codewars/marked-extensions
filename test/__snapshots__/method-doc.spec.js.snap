// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`%method-doc should match snapshot 1`] = `
<h1 id="documentation">
  Documentation
</h1>
<p>
  This is some text
</p>
<div class="block block--docs method-doc">
  <h3>
    <code>
      <dfn class="doc-name doc-name--method">
        open_files
      </dfn>
    </code>
    (
    <code>
      <dfn class="doc-name doc-name--name">
        files
      </dfn>
    </code>
    ,
    <code>
      <dfn class="doc-name doc-name--name">
        ext
      </dfn>
    </code>
    )
  </h3>
  <p>
    Loops through any files that are not opened and opens them
  </p>
  <div class="block block--doc">
    <dl>
      <dt>
        Parameters
      </dt>
      <dd>
        <p>
          <dfn class="doc-name doc-name--name">
            files
          </dfn>
          :
          <dfn class="doc-type">
            Array&lt;File&gt;
          </dfn>
          - An array of file objects that may need to be opened
        </p>
      </dd>
      <dd>
        <p>
          <dfn class="doc-name doc-name--name">
            ext
          </dfn>
          :
          <dfn class="doc-type">
            String
          </dfn>
          - An optional extension that may be used as a filter to determine which files to open
        </p>
      </dd>
      <dt>
        Return Value
      </dt>
      <dd>
        <p>
          <dfn class="doc-type">
            Array&lt;File&gt;
          </dfn>
          - A filtered array of files that were opened
        </p>
      </dd>
      <dt>
        Constraints
      </dt>
      <dd>
        <p>
          0 &lt;= files.length &lt;= 10
        </p>
      </dd>
    </dl>
  </div>
  <div class="block block--doc-block">
    <h4>
      Examples
    </h4>
    <table>
      <thead>
        <tr>
          <th>
          </th>
          <th>
            files
          </th>
          <th>
            ext
          </th>
          <th>
            Return Value
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            <em>
              Example A
            </em>
          </td>
          <td>
            <code>
              [1,2,3]
            </code>
          </td>
          <td>
            <code>
              "js"
            </code>
          </td>
          <td>
            <code>
              [1,2]
            </code>
          </td>
        </tr>
        <tr>
          <td>
            <em>
              Example
            </em>
          </td>
          <td>
            <code>
              [1,2,3]
            </code>
          </td>
          <td>
            <code>
              "rb"
            </code>
          </td>
          <td>
            <code>
              [3]
            </code>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`%method-doc should match snapshot with language: csharp 1`] = `
<h1 id="documentation">
  Documentation
</h1>
<p>
  This is some text
</p>
<div class="block block--docs method-doc">
  <h3>
    <code>
      <dfn class="doc-class">
        Challenge
      </dfn>
      .
      <dfn class="doc-name doc-name--method">
        OpenFiles
      </dfn>
    </code>
    (
    <code>
      <dfn class="doc-name doc-name--name">
        files
      </dfn>
    </code>
    ,
    <code>
      <dfn class="doc-name doc-name--name">
        ext
      </dfn>
    </code>
    )
  </h3>
  <p>
    Loops through any files that are not opened and opens them
  </p>
  <div class="block block--doc">
    <dl>
      <dt>
        Parameters
      </dt>
      <dd>
        <p>
          <dfn class="doc-name doc-name--name">
            files
          </dfn>
          :
          <dfn class="doc-type">
            File[]
          </dfn>
          - An array of file objects that may need to be opened
        </p>
      </dd>
      <dd>
        <p>
          <dfn class="doc-name doc-name--name">
            ext
          </dfn>
          :
          <dfn class="doc-type">
            string
          </dfn>
          - An optional extension that may be used as a filter to determine which files to open
        </p>
      </dd>
      <dt>
        Return Value
      </dt>
      <dd>
        <p>
          <dfn class="doc-type">
            File[]
          </dfn>
          - A filtered array of files that were opened
        </p>
      </dd>
      <dt>
        Constraints
      </dt>
      <dd>
        <p>
          0 &lt;= files.length &lt;= 10
        </p>
      </dd>
    </dl>
  </div>
  <div class="block block--doc-block">
    <h4>
      Examples
    </h4>
    <table>
      <thead>
        <tr>
          <th>
          </th>
          <th>
            files
          </th>
          <th>
            ext
          </th>
          <th>
            Return Value
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            <em>
              Example A
            </em>
          </td>
          <td>
            <code>
              [1,2,3]
            </code>
          </td>
          <td>
            <code>
              "js"
            </code>
          </td>
          <td>
            <code>
              [1,2]
            </code>
          </td>
        </tr>
        <tr>
          <td>
            <em>
              Example
            </em>
          </td>
          <td>
            <code>
              [1,2,3]
            </code>
          </td>
          <td>
            <code>
              "rb"
            </code>
          </td>
          <td>
            <code>
              [3]
            </code>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`%method-doc should match snapshot with language: javascript 1`] = `
<h1 id="documentation">
  Documentation
</h1>
<p>
  This is some text
</p>
<div class="block block--docs method-doc">
  <h3>
    <code>
      <dfn class="doc-name doc-name--method">
        openFiles
      </dfn>
    </code>
    (
    <code>
      <dfn class="doc-name doc-name--name">
        files
      </dfn>
    </code>
    ,
    <code>
      <dfn class="doc-name doc-name--name">
        ext
      </dfn>
    </code>
    )
  </h3>
  <p>
    Loops through any files that are not opened and opens them
  </p>
  <div class="block block--doc">
    <dl>
      <dt>
        Parameters
      </dt>
      <dd>
        <p>
          <dfn class="doc-name doc-name--name">
            files
          </dfn>
          :
          <dfn class="doc-type">
            Array&lt;File&gt;
          </dfn>
          - An array of file objects that may need to be opened
        </p>
      </dd>
      <dd>
        <p>
          <dfn class="doc-name doc-name--name">
            ext
          </dfn>
          :
          <dfn class="doc-type">
            String
          </dfn>
          - An optional extension that may be used as a filter to determine which files to open
        </p>
      </dd>
      <dt>
        Return Value
      </dt>
      <dd>
        <p>
          <dfn class="doc-type">
            Array&lt;File&gt;
          </dfn>
          - A filtered array of files that were opened
        </p>
      </dd>
      <dt>
        Constraints
      </dt>
      <dd>
        <p>
          0 &lt;= files.length &lt;= 10
        </p>
      </dd>
    </dl>
  </div>
  <div class="block block--doc-block">
    <h4>
      Examples
    </h4>
    <table>
      <thead>
        <tr>
          <th>
          </th>
          <th>
            files
          </th>
          <th>
            ext
          </th>
          <th>
            Return Value
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            <em>
              Example A
            </em>
          </td>
          <td>
            <code>
              [1,2,3]
            </code>
          </td>
          <td>
            <code>
              "js"
            </code>
          </td>
          <td>
            <code>
              [1,2]
            </code>
          </td>
        </tr>
        <tr>
          <td>
            <em>
              Example
            </em>
          </td>
          <td>
            <code>
              [1,2,3]
            </code>
          </td>
          <td>
            <code>
              "rb"
            </code>
          </td>
          <td>
            <code>
              [3]
            </code>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
